### PR TITLE
Add support to configure BSSID in Wi-Fi connect command

### PR DESCRIPTION
**Pull request by rado17**
1.  Use getopt API to process the arguments passed with connect command.
2.  Add support for BSSID configuration in connect call.

### CODE REVIEW

**Comment by rado17** [comment]
@sachinthegreen @VivekUppunda @ajayparida @kapbh  Please take a look

File: subsys/net/l2/wifi/wifi_shell.c [hunk 1]
```diff
@@ -442,73 +442,68 @@ static void wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
...
 
-static int __wifi_args_to_params(size_t argc, char *argv[],
+static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv[],
 				 struct wifi_connect_req_params *params,
 				 enum wifi_iface_mode iface_mode)
 {
 	char *endptr;
 	int idx = 1;
-	const struct shell *sh = context.sh;
+	//const struct shell *sh = context.sh;
```
Comment (VivekUppunda) [inline subsys/net/l2/wifi/wifi_shell.c:451]: This can probably be removed if it is not needed?

File: subsys/net/l2/wifi/wifi_shell.c [hunk 2]
```diff
@@ -537,54 +532,44 @@ static int __wifi_args_to_params(size_t argc, char *argv[],
...
-		/* Security type (optional) */
-		if (idx < argc) {
-			unsigned int security = strtol(argv[idx], &endptr, 10);
-
-			if (security <= WIFI_SECURITY_TYPE_MAX) {
-				params->security = security;
+			break;
+		case 'b':
+			if (iface_mode == WIFI_MODE_INFRA) {
+					/* Negative channel means band */
```
Comment (VivekUppunda) [inline subsys/net/l2/wifi/wifi_shell.c:538]: is this conveying the correct statement? Please re-frame this accordingly

File: subsys/net/l2/wifi/wifi_shell.c [hunk 3]
```diff
@@ -442,73 +442,67 @@ static void wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
 	}
```
Comment (jukkar) [inline subsys/net/l2/wifi/wifi_shell.c:442]: Please change the commit subject to `net: wifi: shell: ...` for both commits.

Also add information in the commit message why this change is needed, it might make sense to add information about this change to 3.7 migration guide as it changes user seeing interface.

File: subsys/net/l2/wifi/wifi_shell.c [hunk 4]
```diff
@@ -442,73 +442,67 @@ static void wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
...
+					       {"band", required_argument, 0, 'b'},
+					       {"channel", required_argument, 0, 'c'},
+					       {"help", no_argument, 0, 'h'},
+					       {0, 0, 0, 0}};
+	int opt_index = 0;
+	uint8_t band;
+	long channel = strtol(argv[idx], &endptr, 10);
+	const uint8_t all_bands[] = {WIFI_FREQ_BAND_2_4_GHZ,
+					WIFI_FREQ_BAND_5_GHZ,
+					WIFI_FREQ_BAND_6_GHZ};
```
Comment (jukkar) [inline subsys/net/l2/wifi/wifi_shell.c:468]: Could you modify this a bit to something like this
```
const uint8_t all_bands[] = {
	WIFI_FREQ_BAND_2_4_GHZ,
	WIFI_FREQ_BAND_5_GHZ,
	WIFI_FREQ_BAND_6_GHZ
};
```

File: subsys/net/l2/wifi/wifi_shell.c (hunk 1 above)
Comment (rado17) [reply subsys/net/l2/wifi/wifi_shell.c]: Done

File: subsys/net/l2/wifi/wifi_shell.c (hunk 3 above)
Comment (rado17) [reply subsys/net/l2/wifi/wifi_shell.c:442]: Done

File: subsys/net/l2/wifi/wifi_shell.c (hunk 4 above)
Comment (rado17) [reply subsys/net/l2/wifi/wifi_shell.c]: Done

File: doc/releases/migration-guide-3.7.rst [hunk 5]
```diff
@@ -170,6 +170,13 @@ Networking
   one IPv4 address / network interface, the netmask must be specified
```
Comment (jukkar) [inline doc/releases/migration-guide-3.7.rst:170]: Please change the migration guide commit subject to
```
 doc: migration-guide-3.7: Add wifi shell connect cmd info
```
because we are actually changing the documentation and not the wifi stuff in this commit.

File: doc/releases/migration-guide-3.7.rst [hunk 6]
```diff
@@ -170,6 +170,13 @@ Networking
...
   for each IPv4 address separately. (:github:`68419`)
 
+* Modified the `wifi connect` command to use key-value format for the arguments. In
+  previous implementation, we were identifying the option by option's position in argument string.
+  This makes it difficult to deal with optional arguments or extending the support
+  for other options. Having key-value format makes it easier to extend the options that
+  can be passed to connect command.
+  `wifi -h` will give more information about the usage of connect command.
```
Comment (jukkar) [inline doc/releases/migration-guide-3.7.rst:178]: Could you add the link to this PR like in other bullet points, so
```
(:github:`70024`)
```

**Review by jukkar — COMMENTED** [review]
LGTM, just a minor nit for the migration commit.

File: doc/releases/migration-guide-3.7.rst (hunk 5 above)
Comment (rado17) [reply doc/releases/migration-guide-3.7.rst:170]: Done

File: doc/releases/migration-guide-3.7.rst (hunk 6 above)
Comment (rado17) [reply doc/releases/migration-guide-3.7.rst:178]: Done.

**Review by jukkar — DISMISSED** [review]

**Review by sachinthegreen — APPROVED** [review]

**Review by jukkar — DISMISSED** [review]

File: doc/releases/migration-guide-3.7.rst [hunk 7]
```diff
@@ -175,6 +175,14 @@ Networking
   one IPv4 address / network interface, the netmask must be specified
   for each IPv4 address separately. (:github:`68419`)
 
+* Modified the `wifi connect` command to use key-value format for the arguments. In
```
Comment (kartben) [inline doc/releases/migration-guide-3.7.rst:178]:
```suggestion
* Modified the ``wifi connect`` command to use key-value format for the arguments. In
```

File: doc/releases/migration-guide-3.7.rst [hunk 8]
```diff
@@ -175,6 +175,14 @@ Networking
...
+  previous implementation, we were identifying the option by option's position in argument string.
```
Comment (kartben) [inline doc/releases/migration-guide-3.7.rst:179]:
```suggestion
  previous implementation, we were identifying the option using its position in argument string.
```

File: doc/releases/migration-guide-3.7.rst [hunk 9]
```diff
@@ -175,6 +175,14 @@ Networking
...
+  This makes it difficult to deal with optional arguments or extending the support
+  for other options. Having key-value format makes it easier to extend the options that
+  can be passed to connect command.
+  `wifi -h` will give more information about the usage of connect command.
```
Comment (kartben) [inline doc/releases/migration-guide-3.7.rst:183]:
```suggestion
  ``wifi -h`` will give more information about the usage of connect command.
```

File: subsys/net/l2/wifi/wifi_shell.c [hunk 10]
```diff
@@ -442,73 +442,69 @@ static void wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
...
+			if (params->security) {
+				secure_connection = true;
+			}
+			break;
+		case 'p':
+			if (secure_connection) {
+				params->psk = optarg;
+				params->psk_length = strlen(params->psk);
+			} else {
+				PR_WARNING("Password configured without security configuration\n");
```
Comment (kartben) [inline subsys/net/l2/wifi/wifi_shell.c:503]: nit, perhaps:

```suggestion
				PR_WARNING("Password provided without security configuration\n");
```

Comment (kartben) [reply subsys/net/l2/wifi/wifi_shell.c:503]: also, you seem to be using "passphrase" / "password" inconsistently

**Review by kartben — DISMISSED** [review]

File: doc/releases/migration-guide-3.7.rst (hunk 7 above)
Comment (rado17) [reply doc/releases/migration-guide-3.7.rst]: Done

File: doc/releases/migration-guide-3.7.rst (hunk 8 above)
Comment (rado17) [reply doc/releases/migration-guide-3.7.rst]: Done.

File: doc/releases/migration-guide-3.7.rst (hunk 9 above)
Comment (rado17) [reply doc/releases/migration-guide-3.7.rst]: Done

File: subsys/net/l2/wifi/wifi_shell.c (hunk 10 above)
Comment (rado17) [reply subsys/net/l2/wifi/wifi_shell.c]: Updated the log to use `Passphrase` instead of `Password`.

File: subsys/net/l2/wifi/wifi_shell.c [hunk 11]
```diff
@@ -500,7 +500,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 				params->psk = optarg;
 				params->psk_length = strlen(params->psk);
 			} else {
-				PR_WARNING("Password configured without security configuration\n");
+				PR_WARNING("Passphrase provided without security configuration\n");
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:503]: This commit can be squashed in to the first one where the change was introduced.

File: subsys/net/l2/wifi/wifi_shell.c [hunk 12]
```diff
@@ -1858,9 +1862,11 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		  "[-k, --key-mgmt]: Key Management type (valid only for secure SSIDs)\n"
 		  "0:None, 1:WPA2-PSK, 2:WPA2-PSK-256, 3:SAE, 4:WAPI, 5:EAP, 6:WEP, 7: WPA-PSK\n"
 		  "[-w, --ieee-80211w]: MFP (optional: needs security type to be specified)\n"
-		  ": 0:Disable, 1:Optional, 2:Required.\n",
+		  ": 0:Disable, 1:Optional, 2:Required.\n"
+		  "[-m, --bssid]: BSSID of the AP.\n"
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:1866]: `MAC address of the AP (BSSID)` at least maps to `-m`

File: include/zephyr/net/wifi_mgmt.h [hunk 13]
```diff
@@ -359,6 +359,8 @@ struct wifi_connect_req_params {
 	enum wifi_security_type security;
 	/** MFP options */
 	enum wifi_mfp_options mfp;
+	/** BSSID */
+	uint8_t *bssid;
```
Comment (krish2718) [inline include/zephyr/net/wifi_mgmt.h:363]: should we make this an array as the size is known at compile time?

File: doc/releases/migration-guide-3.7.rst [hunk 14]
```diff
@@ -175,6 +175,14 @@ Networking
   one IPv4 address / network interface, the netmask must be specified
   for each IPv4 address separately. (:github:`68419`)
 
+* Modified the ``wifi connect`` command to use key-value format for the arguments. In
+  previous implementation, we were identifying the option using it's position in argument string.
```
Comment (kartben) [inline doc/releases/migration-guide-3.7.rst:179]: looks like the change I previously suggested wasn't correctly applied

```suggestion
  previous implementation, we were identifying the option using its position in argument string.
```

File: include/zephyr/net/wifi_mgmt.h (hunk 13 above)
Comment (rado17) [reply include/zephyr/net/wifi_mgmt.h:363]: This way we avoid an extra `memcpy` which may be required if we use a static array.

Comment (krish2718) [reply include/zephyr/net/wifi_mgmt.h:363]: It might causes issues with memory management, no? In `wifi_shell.c` you are passing `optarg`, and once the shell function returns it will be freed (as `argv` will be out of scope).

File: doc/releases/migration-guide-3.7.rst [hunk 15]
```diff
@@ -170,6 +170,14 @@ Networking
   one IPv4 address / network interface, the netmask must be specified
   for each IPv4 address separately. (:github:`68419`)
 
+* Modified the `wifi connect` command to use key-value format for the arguments. In
```
[the remark this answers is not in the document]
Comment (rado17) [reply doc/releases/migration-guide-3.7.rst]: Done.

File: include/zephyr/net/wifi_mgmt.h (hunk 13 above)
Comment (rado17) [reply include/zephyr/net/wifi_mgmt.h]: Done.

File: subsys/net/l2/wifi/wifi_shell.c (hunk 12 above)
Comment (rado17) [reply subsys/net/l2/wifi/wifi_shell.c]: Done

File: subsys/net/l2/wifi/wifi_shell.c (hunk 11 above)
Comment (rado17) [reply subsys/net/l2/wifi/wifi_shell.c:503]: Done

**Review by krish2718 — DISMISSED** [review]

**Review by rlubos — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@rado17 this needs a rebase

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]

File: subsys/net/l2/wifi/wifi_shell.c [hunk 16]
```diff
@@ -442,73 +442,69 @@ static void wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
...
+			break;
+		case 'p':
+			if (secure_connection) {
+				params->psk = optarg;
+				params->psk_length = strlen(params->psk);
+			} else {
+				PR_WARNING("Passphrase provided without security configuration\n");
+			}
+			break;
+		case 'c':
```
Comment (WilliamGFish) [inline subsys/net/l2/wifi/wifi_shell.c:506]: /subsys/net/l2/wifi/wifi_shell.c:507:4: error: a label can only be part of a statement and a declaration is not a statement
  507 |    long channel = strtol(optarg, &endptr, 10);
